analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs

```diff
@@ -18,19 +18,37 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System.Linq;
+using Microsoft.CodeAnalysis;
 using SonarAnalyzer.SymbolicExecution.Constraints;
 using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.OperationProcessors
 {
     internal static class Invocation
     {
-        public static ProgramState Process(SymbolicContext context, IInvocationOperationWrapper invocation) =>
-            !invocation.TargetMethod.IsStatic               // Also applies to C# extensions
-            && !invocation.TargetMethod.IsExtensionMethod   // VB extensions in modules are not marked as static
-            && invocation.Instance.TrackedSymbol() is { } symbol
-                ? context.SetSymbolConstraint(symbol, ObjectConstraint.NotNull)
-                : context.State;
+        public static ProgramState Process(SymbolicContext context, IInvocationOperationWrapper invocation)
+        {
+            var state = context.State;
+            if (!invocation.TargetMethod.IsStatic             // Also applies to C# extensions
+                && !invocation.TargetMethod.IsExtensionMethod // VB extensions in modules are not marked as static
+                && invocation.Instance.TrackedSymbol() is { } symbol)
+            {
+                state = state.SetSymbolConstraint(symbol, ObjectConstraint.NotNull);
+            }
+            if (invocation.Instance.UnwrapConversion() is { Kind: OperationKindEx.InstanceReference }
+                || (invocation is { TargetMethod.IsExtensionMethod: true, Arguments: { Length: > 0 } arguments }
+                    && arguments[0] is { Kind: OperationKindEx.Argument }
+                    && IArgumentOperationWrapper.FromOperation(arguments[0]).Value.UnwrapConversion() is { Kind: OperationKindEx.InstanceReference }))
+            {
+                // Invocations on "this" removes IsNull constraints on fields. We assume the fields initialized after a call.
+                foreach (var field in state.SymbolsWith(ObjectConstraint.Null).OfType<IFieldSymbol>().Where(x => !x.IsStatic))
+                {
+                    state = state.SetSymbolValue(field, state[field].WithoutConstraint(ObjectConstraint.Null));
+                }
```

Comment:
I decided to remove only the `ObjectConstraint.IsNull` from the fields. This is the most minimal scope to fix the cases in `NullPointerDereference.cs`. It also seems the most sensible thing to do. If you agree, I would
* Expand this to properties
* Add test cases, to make sure only `ObjectCconstraint.Null` is removed

Reply:
We need more universal solution as dicussed previously. See the old constraints here:
https://github.com/SonarSource/sonar-dotnet/tree/master/analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Sonar/Constraints

Most of them will need to be removed too. Basically only some (SaltSize, Serialization?, InitializationVector, Disposable) should remain.

If we forget `Null`, we have to also forget `NotNull`. Otherwise we'll end up in weird scenarios due to `condition always evaluates to true` behavior